node/network/approval-distribution/src/lib.rs

```diff
@@ -1637,20 +1653,89 @@ async fn adjust_required_routing_and_propagate<Context, BlockFilter, RoutingModi
 	}
 }
 
+#[derive(Debug, Clone)]
+struct ReputationAggregator {
+	overflow: bool,
+	by_peer: std::collections::HashMap<PeerId, i32>,
+}
+
+impl ReputationAggregator {
+	pub fn new() -> Self {
+		Self { overflow: false, by_peer: std::collections::HashMap::new() }
+	}
+
+	pub fn clear(&mut self) {
+		self.by_peer.clear();
+	}
+
+	pub fn update(&mut self, peer_id: PeerId, rep: Rep) {
+		if matches!(rep, Rep::Malicious(_)) {
+			self.overflow = true;
+		}
+		let current = match self.by_peer.get(&peer_id) {
+			Some(v) => *v,
+			None => 0,
+		};
+		let new_value = current.saturating_add(rep.cost_or_benefit());
+		self.by_peer.insert(peer_id, new_value);
+	}
+
+	pub fn overflow(&self) -> bool {
+		self.overflow
+	}
+
+	pub fn by_peer(&self) -> &std::collections::HashMap<PeerId, i32> {
+		&self.by_peer
+	}
+}
+
 /// Modify the reputation of a peer based on its behavior.
-async fn modify_reputation(
-	sender: &mut impl overseer::ApprovalDistributionSenderTrait,
-	peer_id: PeerId,
-	rep: Rep,
-) {
+async fn modify_reputation(reputation: &mut ReputationAggregator, peer_id: PeerId, rep: Rep) {
 	gum::trace!(
 		target: LOG_TARGET,
 		reputation = ?rep,
 		?peer_id,
 		"Reputation change for peer",
 	);
 
-	sender.send_message(NetworkBridgeTxMessage::ReportPeer(peer_id, rep)).await;
+	reputation.update(peer_id, rep);
+}
+
+async fn maybe_send_reputation(
+	sender: &mut impl overseer::ApprovalDistributionSenderTrait,
+	reputation: &mut ReputationAggregator,
+	mut reputation_delay: &mut Fuse<futures_timer::Delay>,
+	reputation_interval: &std::time::Duration,
+) -> bool {
+	if reputation_interval.is_zero() || reputation.overflow() {
+		send_reputation(sender, reputation).await;
+		return true
+	}
+
+	select! {
+		_ = reputation_delay => {
+			send_reputation(sender, reputation).await;
+			true
+		},
+		default => false, // Will run if no futures are immediately ready
+	}
+}
+
+async fn send_reputation(
+	sender: &mut impl overseer::ApprovalDistributionSenderTrait,
+	reputation: &ReputationAggregator,
+) {
+	for (&peer_id, &score) in reputation.by_peer() {
+		sender
+			.send_message(NetworkBridgeTxMessage::ReportPeer(
```

Comment:
We could also consider sending the hashmap directly to the network bridge in a new message type ReportPeers, to avoid the overhead of sending many small messages.

Reply:
Yep.